src/DocParser.php

```diff
@@ -132,15 +133,19 @@ private function incorporateLine(ContextInterface $context)
         }
     }
 
-    private function processInlines(ContextInterface $context, AbstractBlock $block)
+    private function processInlines(ContextInterface $context, NodeWalker $walker)
     {
-        if ($block instanceof AbstractInlineContainer) {
-            $cursor = new Cursor(trim($block->getStringContent()));
-            $block->setInlines($this->inlineParserEngine->parse($context, $cursor));
-        }
+        while (($event = $walker->next()) !== null) {
+            if (!$event->isEntering()) {
+                continue;
+            }
 
-        foreach ($block->getChildren() as $child) {
-            $this->processInlines($context, $child);
+            $node = $event->getNode();
+            if ($node instanceof InlineContainer) {
+                $context->setContainer($node);
+                $cursor = new Cursor(trim($node->getStringContent()));
+                $this->inlineParserEngine->parse($context, $cursor);
```

Comment:
Prior to this PR, the **only** reason we pass the `$context` in is because [`CloseBrackParser` needs access to the document's reference map](https://github.com/thephpleague/commonmark/blob/master/src/Inline/Parser/CloseBracketParser.php#L84).  This seems like a bad architectural decision on my part - we should only pass along what we need, and we certainly don't need the whole `$context` here.  (The `Context` class is primarily designed for tracking the state while handling *blocks*, whereas `InlineParserContext` is designed to track state when dealing with *inlines*.)

Your PR removes `$inlineContext->getInlines()` and uses `$context->getContainer()` instead.  This makes us even more reliant on the `$context`, which I'd like to avoid if possible.

Therefore, if we wanted to avoid using the block-oriented `$context` when parsing inlines, we'd need to make the following changes:

## 1. Add new properties to the `InlineParserContext`

We'll want to add `Node $container` and a `ReferenceMap $refMap` to the constructor plus the corresponding properties and getter methods.

## 2. Rewrite the `InlineParserEngine::parse` method

Change the method from this:

~~~php
 /**
     * @param ContextInterface $context
     * @param Cursor           $cursor
     *
     * @return ArrayCollection
     */
    public function parse(ContextInterface $context, Cursor $cursor)
    {
        $inlineParserContext = new InlineParserContext($cursor);
        /// ....
    }
~~~

To this:

~~~
 /**
     * @param Node         $container
     * @param ReferenceMap $refMap
     *
     * @return ArrayCollection
     */
    public function parse(Node $container, ReferenceMap $refMap)
    {
        $inlineParserContext = new InlineParserContext($node, $refMap);
        // ...
    }
~~~

And have the `InlineParserContext` create the `Cursor` it needs from the node, instead of [relying on an external caller to create that](https://github.com/thephpleague/commonmark/pull/169/files#diff-64d43afca3fac22bce1c4d9a3b84310aR146).

## 3. Remove the `ContextInterface $context` parameter from `AbstractInlineParser::parse()` and its subclasses

The method would change from:

~~~php
public function parse(ContextInterface $context, InlineParserContext $inlineContext)
~~~

To:

~~~php
public function parse(InlineParserContext $inlineContext)
~~~

The container would be obtained via `$inlineContext->getContainer()` and the reference map would be obtained via `$inlineContext->getReferenceMap()`.

## 4. Update the above `DocParser::processInlines` method

The code inside  the `while` loop would instead look like this:

~~~php
$node = $event->getNode();
if ($node instanceof InlineContainer) {
    $this->inlineParserEngine->parse($node);
}
~~~

----

I think this approach greatly simplifies and decouples things.  What are your thoughts?

If you like this approach, would you like to implement these changes or would you prefer that I do it?  Either is fine with me - just let me know :)